### ISSUE TITLE
Support ignoring a branch only on main repo

### ISSUE THREAD

**Issue by nemequ** [travis-ci/travis-ci#6632]
For one of my projects ([Squash](https://quixdb.github.io/squash/)), I only accept patches to master after they have gone through CI (it's a protected branch on GitHub).

Since the commits have all been verified before pushing to master having Travis build them again is a waste, so I've been ignoring master in my .travis.yml.  Unfortunately, this also causes Travis to ignore pull requests coming from a fork's master branch, which means people who submit PRs from their master branch don't get built.

I don't want to remove the exception for master… we have a pretty big build matrix and a relatively large project, and I don't want to waste your resources (especially since you're providing the service for free), but AFAICT there isn't a way to ignore commits to the master branch *only* on a specific repository.  It would be great if you could implement something like that…

**Comment by BanzaiMan** [comment]
I don't think this is how it works. When a fork opens a PR, the PR build's branch is set to the target, not the fork's branch.

I forked your repository and opened 2 PRs https://github.com/quixdb/squash/pull/213 and https://github.com/quixdb/squash/pull/214. The former is from my fork's `master`, and the latter from my fork's `topic-branch`. Neither PR builds run, because both are for the `master` branch, which is forbidden.

**Comment by BanzaiMan** [comment]
When the PR removes `master` from the list, though, it will run (https://travis-ci.org/quixdb/squash/builds/164202397) https://github.com/quixdb/squash/pull/215, but you can reject the PR.

**Comment by BanzaiMan** [comment]
I feel that something is missing from your description of the issue at hand.

**Comment by nemequ** [comment]
Sorry, I got busy and forgot to respond to this until someone submitted another PR.

> I feel that something is missing from your description of the issue at hand.

Okay, I'll try to explain it a bit more clearly.

First off, we have a pretty big build matrix with lots of different versions of gcc and clang, an icc build, AddressSanitizer, coverage analysis, etc.  While Squash isn't a huge project, it's large enough that each build takes a couple minutes on Travis.  This adds up to each build consuming a decent amount of Travis' resources.  Travis is offering this service for free, and we're arguably already abusing it, so I'd really like to avoid extraneous builds.

I like for the master branch to always be buildable, so I've configured it as a [protected branch](https://help.github.com/articles/about-protected-branches/), which means *nothing* gets pushed to master until that exact commit has been verified as working.  For myself and collaborators this usually means pushing to a branch first (generally "staging" or a feature branch like "wip/foo"), then merging into master once the checks have passed.

The problem with this is that by default, when we push something to master Travis will kick off a build, even if that exact commit (same commit ID) has already been checked.  To get around this, I tweaked the .travis.yml to ignore the master branch; after all, GitHub won't let anyone push to it until checks pass on some other branch or PR.  A second build is redundant, and simply wastes Travis' resources.

The problem is that now pull requests targeting the master branch don't work.  As you said, the PR's branch is set to the target, so this makes sense.

What I want is a way to avoid Travis kicking off a redundant build.  If a build works in staging, and I push that exact same commit to master Travis shouldn't kick off another build.  However, pull requests targeting master should run.

**Comment by BanzaiMan** [comment]
> What I want is a way to avoid Travis kicking off a redundant build. If a build works in staging, and I push that exact same commit to master Travis shouldn't kick off another build. However, pull requests targeting master should run.

I think this is what https://github.com/travis-ci/travis-ci/issues/2200 aims to do.

**Comment by nemequ** [comment]
Hm.  It would fix this, though I immediately had the same concern about Coverity Scan… maybe you could consider this an alternate proposal with different trade-offs?  That proposal would have the benefit of working automatically, but this one would provide more flexibility.
